File: threedollar-domain/src/main/java/com/depromeet/threedollar/domain/boss/domain/account/repository/BossAccountRepositoryCustomImpl.java
package com.depromeet.threedollar.domain.boss.domain.account.repository;

import com.depromeet.threedollar.domain.boss.domain.account.BossAccount;
import com.depromeet.threedollar.domain.boss.domain.account.BossAccountSocialType;
import com.querydsl.jpa.impl.JPAQueryFactory;
import lombok.RequiredArgsConstructor;

import static com.depromeet.threedollar.domain.boss.domain.account.QBossAccount.*;

@RequiredArgsConstructor
public class BossAccountRepositoryCustomImpl implements BossAccountRepositoryCustom {

    private final JPAQueryFactory queryFactory;

    @Override
    public BossAccount findBossAccountBySocialTypeAndSocialId(String socialId, BossAccountSocialType socialType) {
        return queryFactory.selectFrom(bossAccount)
            .where(
                bossAccount.socialInfo.socialId.eq(socialId),
                bossAccount.socialInfo.socialType.eq(socialType)
            ).fetchOne();
    }

    @Override
    public BossAccount findBossAccountById(Long bossAccountId) {
        return queryFactory.selectFrom(bossAccount)
            .where(
                bossAccount.id.eq(bossAccountId)
            ).fetchOne();
    }

}

File: threedollar-domain/threedollar-domain-mongo/src/main/kotlin/com/depromeet/threedollar/domain/mongo/domain/bossservice/account/repository/BossAccountRepositoryCustomImpl.kt
package com.depromeet.threedollar.domain.mongo.domain.bossservice.account.repository

import com.depromeet.threedollar.domain.mongo.domain.bossservice.account.BossAccount
import com.depromeet.threedollar.domain.mongo.domain.bossservice.account.BossAccountSocialType
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.data.mongodb.core.findOne
import org.springframework.data.mongodb.core.query.Criteria
import org.springframework.data.mongodb.core.query.Query
import org.springframework.data.mongodb.core.query.isEqualTo
import org.springframework.stereotype.Repository

@Repository
class BossAccountRepositoryCustomImpl(
    private val mongoTemplate: MongoTemplate,
) : BossAccountRepositoryCustom {

    override fun existsBossAccountById(id: String): Boolean {
        return mongoTemplate.exists(Query()
            .addCriteria(BossAccount::id isEqualTo id), BossAccount::class.java
        )
    }

    override fun findBossAccountById(id: String): BossAccount? {
        return mongoTemplate.findOne(Query()
            .addCriteria(BossAccount::id isEqualTo id)
        )
    }

    override fun findBossAccountBySocialInfo(socialId: String, socialType: BossAccountSocialType): BossAccount? {
        return mongoTemplate.findOne(Query()
            .addCriteria(Criteria.where("socialInfo.socialId").isEqualTo(socialId))
            .addCriteria(Criteria.where("socialInfo.socialType").isEqualTo(socialType))
        )
    }

    override fun existsBossAccountBySocialInfo(socialId: String, socialType: BossAccountSocialType): Boolean {
        return mongoTemplate.exists(Query()
            .addCriteria(Criteria.where("socialInfo.socialId").isEqualTo(socialId))
            .addCriteria(Criteria.where("socialInfo.socialType").isEqualTo(socialType)), BossAccount::class.java
        )
    }

}
